Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,1481 @@
+instance zero : Zero (SkewMonoidAlgebra k G) := ⟨⟨0⟩⟩
+
+instance add' : Add (SkewMonoidAlgebra k G) := ⟨add⟩
+
+instance smulZeroClass {S : Type*} [SMulZeroClass S k] :
+    SMulZeroClass S (SkewMonoidAlgebra k G) where
+  smul r p := ⟨r • p.toFinsupp⟩
+  smul_zero a := congr_arg ofFinsupp (smul_zero a)
+
+@[simp]
+theorem ofFinsupp_zero : (⟨0⟩ : SkewMonoidAlgebra k G) = 0 := rfl
+
+@[simp]
+theorem ofFinsupp_add {a b} : (⟨a + b⟩ : SkewMonoidAlgebra k G) = ⟨a⟩ + ⟨b⟩ :=
+  show _ = add _ _ by rw [add_def]
+
+@[simp]
+theorem ofFinsupp_smul {S : Type*} [SMulZeroClass S k] (a : S) (b) :
+    (⟨a • b⟩ : SkewMonoidAlgebra k G) = (a • ⟨b⟩ : SkewMonoidAlgebra k G) :=
+  rfl
+
+@[simp]
+theorem toFinsupp_zero : (0 : SkewMonoidAlgebra k G).toFinsupp = 0 := rfl
+
+@[simp]
+theorem toFinsupp_add (a b : SkewMonoidAlgebra k G) :
+    (a + b).toFinsupp = a.toFinsupp + b.toFinsupp := by
+  cases a
+  cases b
+  rw [← ofFinsupp_add]
+
+@[simp]
+theorem toFinsupp_smul {S : Type*} [SMulZeroClass S k] (a : S) (b : SkewMonoidAlgebra k G) :
+    (a • b).toFinsupp = a • b.toFinsupp :=
+  rfl
+
+theorem _root_.IsSMulRegular.skewMonoidAlgebra {S : Type*} [Monoid S] [DistribMulAction S k] {a : S}
+    (ha : IsSMulRegular k a) : IsSMulRegular (SkewMonoidAlgebra k G) a
+  | ⟨_x⟩, ⟨_y⟩, h => congr_arg _ <| ha.finsupp (ofFinsupp.inj h)
+
+theorem toFinsupp_injective :
+    Function.Injective (toFinsupp : SkewMonoidAlgebra k G → Finsupp _ _) :=
+  fun ⟨_x⟩ ⟨_y⟩ => congr_arg _
+
+@[simp]
+theorem toFinsupp_inj {a b : SkewMonoidAlgebra k G} : a.toFinsupp = b.toFinsupp ↔ a = b :=
+  toFinsupp_injective.eq_iff
+
+theorem ofFinsupp_injective :
+    Function.Injective (ofFinsupp : Finsupp _ _ → SkewMonoidAlgebra k G) :=
+  fun _x _y => congr_arg toFinsupp
+
+@[simp]
+theorem toFinsupp_eq_zero {a : SkewMonoidAlgebra k G} : a.toFinsupp = 0 ↔ a = 0 := by
+  rw [← toFinsupp_zero, toFinsupp_inj]
+
+/-- A more convenient spelling of `SkewMonoidAlgebra.ofFinsupp.injEq` in terms of `Iff`. -/
+theorem ofFinsupp_inj {a b} : (⟨a⟩ : SkewMonoidAlgebra k G) = ⟨b⟩ ↔ a = b :=
+  iff_of_eq (ofFinsupp.injEq _ _)
+
+@[simp]
+theorem ofFinsupp_eq_zero {a} : (⟨a⟩ : SkewMonoidAlgebra k G) = 0 ↔ a = 0 := by
+  rw [← ofFinsupp_zero, ofFinsupp_inj]
+
+instance inhabited : Inhabited (SkewMonoidAlgebra k G) := ⟨0⟩
+
+instance nontrivial [Nontrivial k] [Nonempty G] :
+    Nontrivial (SkewMonoidAlgebra k G) := Function.Injective.nontrivial ofFinsupp_injective
+
+instance addCommMonoid : AddCommMonoid (SkewMonoidAlgebra k G) :=
+{ Function.Injective.addCommMonoid toFinsupp toFinsupp_injective toFinsupp_zero
+      toFinsupp_add (fun _ _ => toFinsupp_smul _ _) with
+    toAdd  := SkewMonoidAlgebra.add'
+    toZero := SkewMonoidAlgebra.zero
+    nsmul  := (. • .) }
+
+instance isCancelAdd [IsCancelAdd k] : IsCancelAdd (SkewMonoidAlgebra k G) :=
+  Function.Injective.isCancelAdd toFinsupp toFinsupp_injective toFinsupp_add
+
+instance coeFun : CoeFun (SkewMonoidAlgebra k G) fun _ => G → k := ⟨fun s ↦ s.toFinsupp⟩
```

Comment:
Is this not a bit dangerous as an instance? Are we not trying to avoid people thinking of `MonoidAlgebra` as being made of functions?

Reply:
That's a good point, I asked for opinions on Zulip [here](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2310541.3A.20SkewMonoidAlgebra/near/452575660).

Reply:
My impression is that a lot of work in the file consists in transferring by hand results from `Finsupp.` to `SkewMonoidAlgebra.`, in a way that lets users manipulate them from both sides at the same time. 
And maybe this should be specifically prevented.

Reply:
This has been changed.